### PR TITLE
fix(batch): Improve batch verifier async, correctness, and performance

### LINKED ISSUE #4728: Log "batch verification receiver lagged" at info level
## Motivation

This error-level log is alarming to users, but there's nothing the user can do about it:
> 2022-06-30T17:51:02.928651ZERROR{net="Main"}:sync:try_to_sync:extend_tips:zebra_consensus::primitives::redpallas: batch verification receiver lagged and lost verification results

https://github.com/ZcashFoundation/zebra/runs/7135923149?check_suite_focus=true#step:6:644

Each batch verifier has a separate instance of this log, they all need to be fixed.

### LINKED ISSUE #4729: Stop batch verification lagging and failing entire batches
## Motivation

The batch verifier can lag and fail entire blocks, even if the proofs and signatures are valid. We should re-design it so it can't lag.

If we can't do that, the dropped verifications should fail immediately, rather than waiting for the block verification timeout:

```
2022-06-30T17:51:02.943983ZERROR{net="Main"}:sync:try_to_sync:extend_tips:zebra_consensus::primitives::redpallas: batch verification receiver lagged and lost verification results
2022-06-30T17:51:57.423166Z INFO{net="Main"}:zebrad::components::sync::progress: estimated progress to chain tip sync_percent=99.819% current_height=Height(1718380) network_upgrade=Nu5 remaining_sync_blocks=3119 time_since_last_state_block=1m
...
2022-06-30T17:54:57.425405Z INFO{net="Main"}:zebrad::components::sync::progress: estimated progress to chain tip sync_percent=99.819% current_height=Height(1718380) network_upgrade=Nu5 remaining_sync_blocks=3122 time_since_last_state_block=4m
2022-06-30T17:55:57.426619Z INFO{net="Main"}:zebrad::components::sync::progress: estimated progress to chain tip sync_percent=99.819% current_height=Height(1718381) network_upgrade=Nu5 remaining_sync_blocks=3121 time_since_last_state_block=0s
2022-06-30T17:56:47.835664Z WARN{net="Main"}:sync:try_to_sync:zebrad::components::sync: error downloading and verifying block e=Invalid { error: Block(Transaction(InternalDowncastError("downcast to known transaction error type failed, original error: Elapsed(())"))), height: Height(1718445), hash: block::Hash("0000000000f0f61e6f42984784ad367711c0b3e704e840797606314426dc2a90") }
```

https://github.com/ZcashFoundation/zebra/runs/7135923149?check_suite_focus=true#step:6:644

## Designs

We can either:
- replace the batch verifier with a watch channel, and create a new watch channel for each batch
- create a new broadcast channel for each batch, so it only ever has one result in it, and make the channel size 1

### LINKED ISSUE #4738: Check for `Worker` panics in `tower_batch::Batch`
## Motivation

In `Batch::new()`, we spawn a `Worker` task, but drop its `JoinHandle`:
https://github.com/ZcashFoundation/zebra/blob/32faa94fb4151c9dfb7d5efcbee3fd2c1afd4c96/tower-batch/src/service.rs#L69

This means it can panic, and Zebra just continues as if nothing happened. This can create or hide all kinds of bugs, including consensus rule failures.

### Designs

We can check the `JoinHandle` for errors in `Batch::poll_ready`, just like we do for the peer set.

## Related Work

Might be related to:
- #4729

### LINKED ISSUE #4740: Run groth16, ed25519, and redjubjub batches on a blocking thread
## Motivation

In PR #4726, we moved Orchard batches to a blocking thread, to resolve a serious verifier hang bug.

We should do the same thing with the other batch verifiers:
- [ ] groth16 batch flush and verify single
- [ ] ed25519 batch flush and verify single
- [ ] redjubjub batch flush and verify single

### Tokio Docs

CPU-intensive tasks should be run on a blocking thread, so they don't stop other futures from running:
https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html

### Designs

We can just do what we did in PR #4726:
https://github.com/ZcashFoundation/zebra/pull/4726/files#r912670247

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra's tower-batch has some async performance issues, and contains some buggy and risky code.
We should replace it with more robust code.

Close #4738 
Close #4740
Close #4729 (by using a channel that never lags)
Close #4728 (by removing the log entirely)

## API Reference

`tokio::spawn_blocking()`:
https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html

## Solution

Improve tower-batch async fairness:
- Prefer the batch timer if there are also new batch requests
- Allow other tasks to run after each batch
- Label each batch worker with the verifier's type in `tokio-console`

Improve error handling in tower-batch:
- Check batch worker tasks for panics and task termination
- Rename Handle to ErrorHandle, and fix up some docs

Improve channel limit handling in tower-batch:
- Use a new channel for each batch
- Use tokio's PollSemaphore instead of an outdated Semaphore impl

Improve channel limit handling in zebra-consensus:
- Replace batch verifier broadcast channels with watch channels

Run all verifier cryptography on blocking CPU-bound threads
- Run batch verification on blocking `tokio` threads
- Run fallback verification on blocking `tokio` threads
- Make flush and drop behaviour consistent for all verifiers
- Switch batch verifier tests to the multi-threaded runtime

Related fixes:
- Fix a bug in the `v5_with_sapling_spends` test (but it's still missing NU5 test vectors)

## Review

This is blocking `halo2` batch verification, so it would be good to get it in soon.
This might fix some bugs we're seeing during syncing.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] CI passes

### CODE REVIEW

**Comment by teor2345** [comment]
I'm running a full sync here, to see if these fixes help:
https://github.com/ZcashFoundation/zebra/actions/runs/2619584632

**Comment by teor2345** [comment]
I'm running checkpoint and tip syncs here, to check if this PR fixes some sync bugs we're seeing on `main`:
https://github.com/ZcashFoundation/zebra/actions/runs/2646938611

**Comment by teor2345** [comment]
And here's a tip and checkpoint sync without larger batch queues (they might be causing some issues):
https://github.com/ZcashFoundation/zebra/actions/runs/2646961635

**Comment by teor2345** [comment]
Here's a tip and checkpoint sync run, to see if this PR fixes the sync hangs we're seeing on `main`:
(I removed some changes from this PR that were buggy.)

https://github.com/ZcashFoundation/zebra/actions/runs/2646994181

**Comment by teor2345** [comment]
@mergifyio update

File: tower-batch/src/service.rs [hunk 1]
```diff
@@ -128,36 +189,62 @@ where
...
-        // `Poll::Pending`. If it returns Pending, the semaphore also schedules
-        // the task for wakeup when the next permit is available.
-        ready!(self.semaphore.poll_acquire(cx));
+        // Correctness: Registers this task for wakeup when the next permit is available,
+        // or when the semaphore is closed.
+        let permit = ready!(self.semaphore.poll_acquire(cx));
+        if let Some(permit) = permit {
+            // Calling poll_ready() more than once will drop any previous permit,
+            // releasing its capacity back to the semaphore.
+            self.permit = Some(permit);
```
Comment (teor2345) [inline tower-batch/src/service.rs:240]: If the capacity is 1 this could deadlock.

Drop the permit before trying to acquire a new one.

Comment (teor2345) [reply tower-batch/src/service.rs:240]: Fixed in 14d8e7893

**Review by oxarbitrage — COMMENTED** [review]
I think this looks ok but i am not approving yet in case you want to do https://github.com/ZcashFoundation/zebra/pull/4776 on top of this branch first. Feel free to admin merge or let me know how you will want to handle it and ill approve where needed.

**Comment by teor2345** [comment]
I admin-merged this separately from PR #4776, just in case we need to debug them as separate commits.

File: tower-batch/src/service.rs [hunk 2]
```diff
@@ -73,8 +99,28 @@ where
...
+            let batch_kind = std::any::type_name::<T>();
+
+            // TODO: identify the unique part of the type name generically,
+            //       or make it an argument to this method
+            let batch_kind = batch_kind.trim_start_matches("zebra_consensus::primitives::");
+            let batch_kind = batch_kind.trim_end_matches("::Verifier");
+
+            tokio::task::Builder::new()
+                .name(&format!("{} batch", batch_kind))
+                .spawn(worker.run().instrument(span))
```
Comment (str4d) [inline tower-batch/src/service.rs:117]: Bug: five days before this PR was merged, `Builder::spawn` was made fallible in upstream (https://github.com/tokio-rs/tokio/pull/4823). This code has not compiled for the past several years.

**Pull request merged**